src/main/scala/treadle/coverage/pass/AddCoverageExpressions.scala

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package treadle.coverage.pass
+
+import firrtl.Namespace
+import firrtl.PrimOps.Not
+import firrtl.ir.{Block, Circuit, Connect, DefModule, DoPrim, Expression, ExtModule, HasInfo, Module, Mux, NoInfo, Statement}
+import firrtl.passes.Pass
+import firrtl.stage.TransformManager.TransformDependency
+import treadle.coverage.{CoverageInfo, Ledger}
+
+import scala.collection.mutable
+
+/**
+  * Adds additional coverage statements to the Low FIRRTL source at every mux location.
+  * These statements can then be used to gather statement coverage information from inside of the Treadle tester
+  */
+object AddCoverageExpressions extends Pass {
+
+  override def prerequisites: Seq[TransformDependency] = firrtl.stage.Forms.LowForm
+
+  override def name = "Coverage!"
+
+  /**
+    * Run the coverage extension on every module
+    * @param c the circuit on with we want to add the coverage extensions
+    * @return the newly modified version of the circuit including coverage validators
+    */
+  override def run(c: Circuit): Circuit = {
+    val ledger = new Ledger()
+    val newModule =  c.modules.map(coverM(ledger))
```

Comment:
You should probably check that the number of modules is one and throw an exception otherwise since the way that you are currently adding ports will only for for the top level module.
Modules that are lower in the hierarchy would need their coverage ports to somehow be wired to the top-level which is not implemented atm.